### PR TITLE
Added unit test for Issue-4214

### PR DESCRIPTION
**Pull request by dvhvsekhar**
See https://github.com/FasterXML/jackson-databind/issues/4214 for more details.

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java [hunk 1]
```diff
@@ -702,7 +704,10 @@ public void serializeAsField(Object bean, JsonGenerator gen,
         }
         // then find serializer to use
         JsonSerializer<Object> ser = _serializer;
-        if (ser == null) {
+		if (ser == null || (value instanceof EnumSet && ser.getClass() != EnumSetSerializer.class)) {
+			// if value instance of EnumSet, use EnumSetSerializer for EnumSet instead of
+			// generic CollectionSerializer, since de-serialization uses EnumSetDeserializer,
+			// if property value is of type EnumSet
```
Comment (JooHyukKim) [inline src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:710]: I am sorry, but this is not a valid solution. The solution does not seem to fit overall and class design in any way. Is there simliar implementation in `BeanPropertyWriter`? Or any other pattern that you refered to?

File: src/test/java/com/fasterxml/jackson/databind/ser/enums/EnumSetSerializationTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+					return false;
+				}
+				return enumSet.containsAll(eh.getEnumSet());
+			}
+		}
+	}
+
+	public void testSerialization() throws Exception {
+		ObjectMapper mapper = jsonMapperBuilder().build()
+				.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/ser/enums/EnumSetSerializationTest.java:58]: Never disable this for unit tests (except when testing feature itself) -- it has a nasty habit of hiding real problems, and developers wrongly enable it out of habit (there are good reasons to enable it too, just not always).

File: src/test/java/com/fasterxml/jackson/databind/ser/enums/EnumSetSerializationTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+				// this type information is needed in json string, for this test
+				.activateDefaultTyping(BasicPolymorphicTypeValidator.builder().allowIfBaseType(Object.class).build(),
+						DefaultTyping.EVERYTHING);
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/ser/enums/EnumSetSerializationTest.java:61]: Ah. And of course it's `EVERYTHING` that causes problems. :)

**Comment by cowtowncoder** [comment]
@dvhvsekhar Would it be possible to change the PR to just have failing test? It should be moved under `src/test/java/.../failing` so as not to block CI; and name should contain id 4214 to indicate issue it is related with.
With failing test it would be easier for anyone to try a fix. Fix proposed here is unfortunately not acceptable, but test seems valid.

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java (hunk 1 above)
Comment (dvhvsekhar) [reply src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:710]: i agree that the fix is not fitting in overall design of this class, just added this workaround here to explain the problem better. I tried to set the serializer while initializing the instance of this class, but the runtime type information of the property value is not available at the time of initialization this class's object.

File: src/test/java/com/fasterxml/jackson/databind/ser/enums/EnumSetSerializationTest.java (hunk 2 above)
Comment (dvhvsekhar) [reply src/test/java/com/fasterxml/jackson/databind/ser/enums/EnumSetSerializationTest.java:58]: sure, will remove this configuration.

**Comment by dvhvsekhar** [comment]
> should be moved under

sure will do that

**Review by cowtowncoder — APPROVED** [review]

**Pull request merged**